ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java

```diff
@@ -138,8 +141,32 @@ private void finalizeAlterTableWithWriteIdOp(Table table, Table oldTable, List<P
       environmentContext.putToProperties(HiveMetaHook.ALTER_TABLE_OPERATION_TYPE, desc.getType().name());
       if (partitions == null) {
         long writeId = desc.getWriteId() != null ? desc.getWriteId() : 0;
-        context.getDb().alterTable(desc.getDbTableName(), table, desc.isCascade(), environmentContext, true,
-            writeId);
+        try {
+          context.getDb().alterTable(desc.getDbTableName(), table, desc.isCascade(), environmentContext, true, writeId);
+        } catch (HiveException ex) {
+          if (environmentContext.getProperties().containsKey(HiveMetaHook.INITIALIZE_ROLLBACK_ALTER) && Boolean
+              .valueOf(environmentContext.getProperties().get(HiveMetaHook.INITIALIZE_ROLLBACK_ALTER))) {
+            // in case of rollback of alter table do the following:
+            // 1. drop the already altered table but keep the data files
+            // 2. recreate the original table
+            // 3. run msck repair to sync partitions on filesystem with metastore
+            context.getDb().dropTable(table.getDbName(), table.getTableName(), false, true, false);
+            context.getDb().createTable(oldTable);
+            Msck msck = new Msck(false, false);
+            try {
+              msck.init(Msck.getMsckConf(context.getDb().getConf()));
+              msck.updateExpressionProxy(Msck.getProxyClass(context.getDb().getConf()));
+              MsckInfo msckInfo =
+                  new MsckInfo(table.getCatalogName(), table.getDbName(), table.getTableName(), null, null, true, true,
```

Comment:
I wish this `MockInfo` had a builder :) oh well

Reply:
Well, yes :). I was thinking about this, but I would have to change to code in a dozen more places, which doesn't fit into this PR.

Reply:
Yeah, makes sense. I guess we could do it one day when we have way too much time on our hands :)